src/proxy/mod.rs

```diff
@@ -74,58 +88,73 @@ where
         }
     }
 
-    async fn reply_handler(&self, msg: &Message) -> Result<()> {
-        let mut envelope = match self.storage.get(&msg.id).await? {
+    async fn reply_handler(&self, msg: &mut Message) -> Result<()> {
+        let mut envelope = match self.storage.get_envelope(&msg.id).await? {
             Some(envelope) => envelope,
-            None => return Ok(()), //timedout .. nothing to do.
+            None => {
+                // timed-out .. nothing to do.
+                log::debug!("envelope of {} is now expired", &msg.id);
+                return Ok(());
+            }
         };
 
+        // sign the message
+        msg.sign(&self.identity);
+
         envelope.data = base64::encode(msg.to_json().context("failed to encode message")?);
+        // swap the source and destination
+        let source = envelope.source;
+        envelope.source = envelope.destination[0];
+        envelope.destination = vec![source];
+        // send for normal reply
         self.storage.response(&envelope).await
     }
 
-    async fn reply(&self, msg: Message) {
-        if let Err(err) = self.reply_handler(&msg).await {
+    async fn reply(&self, msg: &mut Message) {
```

Comment:
This already was taking  ownership of the message. Why do you need to change this to &mut ? the reply handler then can take also full mut ownership of the message

Reply:
I remember doing this because we do sign the message before sending it for reply.